src/oph/heratepalvelu/tep/kestojenUudelleenlaskentaHandler.clj

```diff
@@ -0,0 +1,68 @@
+(ns oph.heratepalvelu.tep.kestojenUudelleenlaskentaHandler
+  "TODO"
+  (:require [clojure.tools.logging :as log]
+            [environ.core :refer [env]]
+            [oph.heratepalvelu.db.dynamodb :as ddb]
+            [oph.heratepalvelu.tep.tepCommon :as tc]
+            [oph.heratepalvelu.tep.niputusHandler :as nh]
+            [oph.heratepalvelu.log.caller-log :refer
+             [log-caller-details-scheduled]]))
+
+(gen-class :name "oph.heratepalvelu.tep.kestojenUudelleenlaskentaHandler"
+           :methods
+           [[^:static handleKestojenUudelleenlaskenta
+             [com.amazonaws.services.lambda.runtime.events.ScheduledEvent
+              com.amazonaws.services.lambda.runtime.Context] void]])
+
+(def rahoituskausi {:alkupvm "2022-07-01" :loppupvm "2023-06-30"})
+
+(defn scan-for-unhandled-nippus!
+  [exclusive-start-key]
+  (let [resp (ddb/scan
+               {:filter-expression
+                (str
+                  "niputuspvm >= :rahoituskausi_alkupvm AND "
+                  "niputuspvm <= :rahoituskausi_loppupvm AND "
+                  "attribute_not_exists(keston_uudelleenlaskenta_suoritettu)")
+                :exclusive-start-key exclusive-start-key
+                :expr-attr-vals {":rahoituskausi_alkupvm"
+                                 [:s (:alkupvm rahoituskausi)]
+                                 ":rahoituskausi_loppupvm"
+                                 [:s (:loppupvm rahoituskausi)]}}
+               (:nippu-table env))]
+    (log/info "Haettiin onnistuneesti " (count (:items resp)) " nippua")
+    resp))
+
+(defn query-jaksot-with-kesto!
+  "Hakee tietokannasta ne jaksot, jotka kuuluvat parametrina annettuun nippuun.
+  Suodattaan jaksoja edelleen siten, että jäljelle jääviltä jaksoilta löytyy
+  vanhastaan laskettu kesto, mutta ei uudella tavalla laskettua kestoa."
+  [nippu]
+  (ddb/query-items
+    {:ohjaaja_ytunnus_kj_tutkinto
+     [:eq [:s (:ohjaaja_ytunnus_kj_tutkinto nippu)]]
+     :niputuspvm [:eq [:s (:niputuspvm nippu)]]}
+    {:index "niputusIndex"
+     :filter-expression
+     "attribute_exists(kesto) AND attribute_not_exists(kesto_vanha)"}
+    (:jaksotunnus-table env)))
+
+(defn -handleKestojenUudelleenlaskenta
+  [_ event ^com.amazonaws.services.lambda.runtime.Context context]
+  (log-caller-details-scheduled "handleKestojenUudelleenlaskenta" event context)
+  (loop [niput (scan-for-unhandled-nippus! nil)]
+    (doseq [nippu (:items niput)]
+      (let [jaksot (query-jaksot-with-kesto! nippu)
+            kestot (nh/jaksojen-kestot! jaksot)]
```

Comment:
Öööh, siis eikös tässä ole juuri se ongelma, josta eilen mainitsit, eli että kun `jaksojen-kestot!` kutsuu `get-concurrent-jaksot-from-ehoks!` niin sieltä ei tule deletoituja jaksoja joten sieltä saattaa puuttua kokonaan ne jaksot joita varten niitä yhtaikaisia jaksoja kysellään.  Eli jos jakso on päivitetty tässä välissä, sillä on uusi oht_id eikä vanha oht_id löydy tuolta `concurrent-jaksot`-muuttujasta ollenkaan.

Reply:
Tuo on totta. Tuossa tapauksessa, muutama rivi alempana, poistetulle jakso asetetaan kestoksi `0`, mutta tuo ei tietenkään ole hyväksyttävä ratkaisu jos jaksoa on päivitetty eikä varsinaisesti poistettu. :thinking:

Reply:
Niin joo tosiaan, osa jaksoista voi olla aidosti poistuneitakin.

Reply:
Jep, noita kokonaan poistettuja ei osata ottaa uudelleenlaskennassa huomioon muuten kuin asettamalla kestoksi nolla, mutta tuo lienee tässä tilanteessa ihan hyväksyttävää. Olettaen, että tuo eHOKSin `select-tyoelamajaksot-active-between` toimii kuten pitää, pitäisi myös jyvityslaskenta mennä rinnakaisten jaksojen osalta ok kun jakso on poistettu.

Reply:
Täytyy testata näitäkin erikoistapauksia QA:lla.

Reply:
> kun jaksojen-kestot! kutsuu get-concurrent-jaksot-from-ehoks! niin sieltä ei tule deletoituja jaksoja joten sieltä saattaa puuttua kokonaan ne jaksot joita varten niitä yhtaikaisia jaksoja kysellään. Eli jos jakso on päivitetty tässä välissä, sillä on uusi oht_id eikä vanha oht_id löydy tuolta concurrent-jaksot-muuttujasta ollenkaan.

Tämä ongelma tosiaan on nyt korjattu osana tikettiä [EH-1699](https://jira.eduuni.fi/browse/EH-1699).